### PR TITLE
core: add the self-healing feature

### PR DESCRIPTION
**Pull request by taskset**
core: add the self-healing feature

Systemd has registered a crash handler called crash().
When a fatal exception occurs during operation, it will
call the crash () function: collect coredump, and then
call freeze_or_exit_or_reboot().

However, in cloud server scenario, many critical applications
need to minimize downtime. If systemd enters the
freeze_or_exit_or_reboot function, the systemctl command
cannot be executed, and all VMs and containers on the server
may be affected. The server needs to be restarted to recover.

We have added a self-healing feature to systemd, which can
repair systemd exceptions online and minimize downtime of
critical applications on the server.

Signed-off-by: duolong <duolong@alibaba-inc.com>
Signed-off-by: Wen Yang <wenyang@linux.alibaba.com>

### CODE REVIEW

**Comment by boucman** [comment]
So IIUC it's an automatic reexecute of systemd, right ?
Maybe rename it to crash-reexecute (more consistant with crash-reboot and more descriptive)
Also, there is no documentation for this new option in the manpage itself

(I don't have enough knowledge to ACK the patch itself, so I won't comment on the technical merit)

**Comment by w-simon** [comment]
> So IIUC it's an automatic reexecute of systemd, right ?
> Maybe rename it to crash-reexecute (more consistant with crash-reboot and more descriptive)
> Also, there is no documentation for this new option in the manpage itself
> 
> (I don't have enough knowledge to ACK the patch itself, so I won't comment on the technical merit)

Thanks for your comments.
crash-reexecute is indeed more consistent with crash-reboot.
But It is a little different from pure crash-reexecute because it requires manual intervention (analyze the collected vmcore, then upgrade to the systemd version with the bug fixed (rpm -Uvh xxx), and finally trigger the re-execution of the new systemd).

How about we rename it to crash-upgrade-reexec?
Thank you.

File: src/core/main.c [hunk 1]
```diff
@@ -79,6 +79,7 @@
 #include "util.h"
 #include "virt.h"
 #include "watchdog.h"
+#include <setjmp.h>
```
Comment (keszybz) [inline src/core/main.c]: Please put this alphabetically among the system headers at the top (those with `<>`).

File: src/core/main.c [hunk 2]
```diff
@@ -147,9 +149,52 @@ static NUMAPolicy arg_numa_policy;
 /* A copy of the original environment block */
 static char **saved_env = NULL;
 
+static volatile sig_atomic_t flag_self_healing_reexec = 0;
```
Comment (keszybz) [inline src/core/main.c]: We don't say "flag_" in general. Maybe call this "waiting_for_signal_to_reexec" or similar.

File: src/core/main.c [hunk 3]
```diff
@@ -147,9 +149,52 @@ static NUMAPolicy arg_numa_policy;
...
+        struct sigaction sa;
+        sigset_t ss;
+        int r = 0;
+
+        sigemptyset(&ss);
+        sigaddset(&ss, SIGTERM);
+        sigprocmask(SIG_UNBLOCK, &ss, NULL);
+
+        sa.sa_handler = reexec_handler;
+        sa.sa_flags = SA_RESTART;
```
Comment (keszybz) [inline src/core/main.c]: Please just initialize `sa` directly:
```C
const struct sigaction sa = {
   .sa_handler = ...,
   .sd_flags = ...,
};
```

File: src/core/main.c [hunk 4]
```diff
@@ -147,9 +149,52 @@ static NUMAPolicy arg_numa_policy;
...
+        sigaction(SIGTERM, &sa, NULL);
+
+        log_emergency("CrashSelfHealing state entered.");
+        log_emergency("Please analyze the collected coredump and upgrade systemd to the bug fixed version.");
```
Comment (keszybz) [inline src/core/main.c]: Remove "please" from the messages. Nice is nice, but brevity in emergency messages is more important ;)
Maybe "Fatal crash, PID1 halted. Send SIGTERM to PID1 to trigger reexecution."
And then explain about the need to collect the coredump and upgrade systemd in the man page.

File: src/core/main.c [hunk 5]
```diff
@@ -147,9 +149,52 @@ static NUMAPolicy arg_numa_policy;
...
+        log_emergency("After that, please send SIGTERM to trigger the reexecution of the new daemon.");
+        flag_self_healing_reexec = 0;
+        while(1) {
+                usleep(10000);
+                if (flag_self_healing_reexec) {
+                        log_emergency("Exit CrashSelfHealing state.");
```
Comment (keszybz) [inline src/core/main.c]: Here and in other places: make sure that the message is unambiguous as to whether something *should* happen or *is* happening. I.e. "Exiting ...", "Jumping...", etc.

File: src/core/main.c [hunk 6]
```diff
@@ -379,6 +427,14 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
                 else
                         arg_crash_reboot = r;
 
+        } else if (proc_cmdline_key_streq(key, "systemd.crash_self_healing")) {
+
+                r = value ? parse_boolean(value) : true;
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse crash self_healing switch %s, ignoring: %m", value);
```
Comment (keszybz) [inline src/core/main.c]: I see that this copies what the surrounding code does, but the surrounding code is not good ;)
Please use the actual swtich name:
```C
log_warning_errno(r, "Failed to parse %s=%s, ignoring: %m", key, value);
```

File: src/core/main.c [hunk 7]
```diff
@@ -915,6 +974,18 @@ static int parse_argv(int argc, char *argv[]) {
                         }
                         break;
 
+                case ARG_CRASH_SELF_HEALING:
+                        if (!optarg)
+                                arg_crash_self_healing = false;
+                        else {
+                                r = parse_boolean(optarg);
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to parse crash self_healing boolean: \"%s\": %m",
```
Comment (keszybz) [inline src/core/main.c]: Here too, say "parse --crash-..." (whatever the final name is).

File: src/core/main.c [hunk 8]
```diff
@@ -1726,6 +1797,11 @@ static int invoke_main_loop(
         assert(ret_switch_root_init);
         assert(ret_error_message);
 
+        if (sigsetjmp(self_healing_jmp_buf, 1)) {
+                log_emergency("Reexecute now ...");
```
Comment (keszybz) [inline src/core/main.c]: "Will reexecute now...".

**Comment by w-simon** [comment]
> Hi, sorry for the delay in review.
> 
> I think this is an interesting approach. So far the thinking was that on a crash, the program state cannot be trusted, so the only thing that can be done is to either tell the kernel to reboot or to halt. But doing a reexec is much more useful in some scenarios, and I think it would be nice to offer that as an option.
> 
> Let's discuss the naming and high-level operation first:
> The name "self healing" is too vague. "crash-upgrade-reexecute" is better, but it stipulates that an upgrade happened, but the admin might choose to continue without an upgrade, and I'd prefer to name the option based on what it does, and not how we think it should be used. So I think "crash-wait-reexecute" or similar would be better. A plain "crash-reexecute" would be OK too.
> 
> I would be nice if this worked somewhat automatically in the sense that a normal package upgrade would cause the reexec to happen. Or in other words, it would be enough to set the config option to use this, without modifying the packaging or doing any manual steps during upgrade. It turns out that this seems to be already the case — e.g. Fedora does `systemctl daemon-reeexec`, but if that fails it falls back to `kill -TERM 1`, and that would DTRT already.

Thank you very much.
We will follow your comments to improve the code as soon as possible.

**Comment by w-simon** [comment]
updated thanks for the review @keszybz . PTAL.

File: src/core/main.c [hunk 9]
```diff
@@ -147,9 +149,49 @@ static NUMAPolicy arg_numa_policy;
...
+
+        sigaction(SIGTERM, &sa, NULL);
+
+        log_emergency("Fatal crash, PID1 halted. Send SIGTERM to PID1 to trigger reexecution.");
+        waiting_for_signal_to_reexec = 0;
+        while(1) {
+                usleep(10000);
+                if (waiting_for_signal_to_reexec) {
+                        log_emergency("Jumping to manager_reexecute now...");
+                        siglongjmp(crash_wait_reexec_jmp_buf, 1);
```
Comment (poettering) [inline src/core/main.c:183]: hmm, so what happens with any open fds here? they just stay open? that effectively leaks them all

Comment (poettering) [reply src/core/main.c:183]: longjump() is really weird in its semantics, as we'll leak mmaps, fds and everything else here. 

I remember that I played around with execve() from the signal handler to actually reexec ourselves once upon a time, but i couldn't get it to work for pid1.

can you explain why you tried siglongjmp() here? did you play around with just reexecing pid1 properly, i.e. via execve()?

**Comment by poettering** [comment]
Let's close this. I don't think longjmps can work because they leak fds and all kinds of other resources. I'd be open to a patch that tries to self-heal by execve(), so that we start with a fresh process, but the current PR sounds like something that would be very unreliable and not real "heal" anything but just make things worse elsewhere.

execve()'ing on crash is something that the kernel didn#t use to allow, but if it works these days it sounds like an approach we could go, but it would be a new PR I guess, since it would work quite differently. Closing hence...

Hope that makes sense?

**Pull request closed without merging**